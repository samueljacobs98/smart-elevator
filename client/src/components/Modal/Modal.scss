@use "../../assets/scss/variables" as *;

@keyframes fadeInOut {
  0% {
    opacity: 0;
  }
  25% {
    opacity: 1;
  }
  75% {
    opacity: 1;
  }
  100% {
    opacity: 0;
  }
}

.modal {
  position: fixed;
  z-index: 999;
  width: 100%;
  height: 100%;
  background-color: rgba(0, 0, 0, 0.5);
  display: grid;
  place-items: center;

  &--show {
    animation: fadeInOut 4s ease-in-out forwards;
  }

  .modal__content {
    display: grid;
    place-items: center;

    width: fit-content;
    background: $colour-gradient-linear-grey;
    box-shadow: 4px 4px 6px 3px rgba(0, 0, 0, 0.5);

    padding: 1rem;
    margin: auto;
    border-radius: 1rem;

    &__title {
      text-align: center;
    }

    &__main {
      display: flex;
      flex-wrap: nowrap;
      justify-content: center;
      align-items: center;

      &__text {
        font-size: 10rem;
      }
    }
  }
}
