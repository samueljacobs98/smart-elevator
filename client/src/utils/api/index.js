import axios from "axios";

const defaultOnError = (error) => console.error(error);

const baseUrl = process.env.REACT_APP_API_URL || "http://localhost:3001/api";

export const getData = async (path, andThen, onError = defaultOnError) => {
  try {
    const response = await axios.get(`${baseUrl}/${path}`);
    return andThen(response.data);
  } catch (error) {
    // TODO: improve error handling
    onError(error);
  }
};

export const postData = async (
  path,
  data,
  andThen,
  onError = defaultOnError
) => {
  try {
    const response = await axios.post(`${baseUrl}/${path}`, data);
    return andThen(response.data);
  } catch (error) {
    onError(error);
  }
};
