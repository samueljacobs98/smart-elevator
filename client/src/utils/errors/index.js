export class EnvVariableError extends Error {
  constructor(variableName) {
    super(`${variableName} env variable not set or invalid - must be a number`);
  }
}
