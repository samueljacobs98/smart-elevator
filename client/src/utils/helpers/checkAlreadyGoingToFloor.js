const checkAlreadyGoingToFloor = (liftStatus, toFloor) =>
  Object.entries(liftStatus).find(([_, { destinations }]) =>
    destinations.includes(toFloor)
  );

export default checkAlreadyGoingToFloor;
