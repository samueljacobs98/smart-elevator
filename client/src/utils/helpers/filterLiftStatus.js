const filterLiftStatus = (lifts, userFloor) => {
  const filteredData = {};

  Object.keys(lifts).forEach((lift) => {
    if (
      lifts[lift].destinations.includes(userFloor) ||
      lifts[lift].floor === userFloor
    ) {
      filteredData[lift] = lifts[lift];
    }
  });

  return filteredData;
};

export default filterLiftStatus;
