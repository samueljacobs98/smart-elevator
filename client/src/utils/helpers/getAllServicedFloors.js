const getAllServicedFloors = (lifts) =>
  Object.values(lifts).reduce((acc, lift) => {
    lift.serviced_floors.forEach((floor) => {
      if (!acc.includes(floor)) acc.push(floor);
    });

    return acc;
  }, []);

export default getAllServicedFloors;
