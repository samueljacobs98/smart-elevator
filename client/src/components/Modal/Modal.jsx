import { halveArray } from "../../utils/helpers";
import Arrow from "../Arrow/Arrow";
import "./Modal.scss";
import { useEffect } from "react";

const Modal = ({ lift, to, showModal, hideModal, lifts }) => {
  useEffect(() => {
    const timeout = setTimeout(hideModal, 4000);
    return () => clearTimeout(timeout);
  }, [hideModal]);

  const [firstHalf, secondHalf] = halveArray(lifts);

  return (
    <div className={`modal ${showModal ? "modal--show" : ""}`}>
      <div className="modal__content">
        <p className="modal__content__title">The next lift to floor {to}:</p>
        <div className="modal__content__main">
          <Arrow isLeft={true} isActive={firstHalf.includes(lift)} />
          <p className="modal__content__main__text">{lift}</p>
          <Arrow isLeft={false} isActive={secondHalf.includes(lift)} />
        </div>
      </div>
    </div>
  );
};

export default Modal;
