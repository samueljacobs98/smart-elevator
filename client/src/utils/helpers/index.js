import filterLiftStatus from "./filterLiftStatus";
import halveArray from "./halveArray";
import mapLiftStatusData from "./mapLiftStatusData";

export { filterLiftStatus, halveArray, mapLiftStatusData };
