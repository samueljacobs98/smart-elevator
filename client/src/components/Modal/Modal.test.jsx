import { render, screen, act } from "@testing-library/react";
import "@testing-library/jest-dom";
import { halveArray } from "../../utils/helpers";
import Modal from "./Modal";

jest.mock("../../utils/helpers", () => ({
  halveArray: jest.fn(),
}));

const lift = "A";
const to = 5;
const lifts = ["A", "B", "C"];

const mockHideModal = jest.fn();

beforeEach(() => {
  jest.useFakeTimers();
  jest.clearAllMocks();
  halveArray.mockReturnValue([["A", "B"], ["C"]]);
});

describe("Modal", () => {
  const renderComponent = (showModal = false) =>
    render(
      <Modal
        lift={lift}
        to={to}
        showModal={showModal}
        hideModal={mockHideModal}
        lifts={lifts}
      />
    );

  it("renders without crashing", () => {
    renderComponent();

    expect(screen).toBeDefined();
  });

  it("dynamically displays the lift number in the title", () => {
    renderComponent(true);

    const title = screen.getByText(`The next lift to floor ${to}:`);

    expect(title).toBeInTheDocument();
  });

  it("renders the correct lift number", () => {
    renderComponent(true);

    const text = screen.getByText(lift);

    expect(text).toBeInTheDocument();
    expect(text).toHaveClass("modal__content__main__text");
  });

  it("calls hide modal after 4 seconds", () => {
    renderComponent(true);

    act(() => {
      jest.advanceTimersByTime(4000);
    });

    expect(mockHideModal).toHaveBeenCalled();
  });

  it("displays the arrows with the correct states", () => {
    renderComponent(true);

    const leftArrow = screen.getByAltText("Left arrow");
    const rightArrow = screen.getByAltText("Right arrow");

    expect(leftArrow).toHaveClass("arrow arrow--left arrow--active");
    expect(rightArrow).toHaveClass("arrow");
    expect(rightArrow).not.toHaveClass("arrow--left arrow--active");
  });

  it("should match snapshot", () => {
    const { container } = renderComponent(true);

    expect(container).toMatchSnapshot();
  });
});
