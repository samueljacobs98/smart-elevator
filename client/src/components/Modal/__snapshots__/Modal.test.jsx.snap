// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should match snapshot 1`] = `
<div>
  <div
    class="modal modal--show"
  >
    <div
      class="modal__content"
    >
      <p
        class="modal__content__title"
      >
        The next lift to floor 
        5
        :
      </p>
      <div
        class="modal__content__main"
      >
        <img
          alt="Left arrow"
          class="arrow arrow--left arrow--active"
          src="arrow.svg"
        />
        <p
          class="modal__content__main__text"
        >
          A
        </p>
        <img
          alt="Right arrow"
          class="arrow  "
          src="arrow.svg"
        />
      </div>
    </div>
  </div>
</div>
`;
